b2g/components/RemoteControlService.jsm

```diff
@@ -0,0 +1,395 @@
+    let server = this.server;
+    server._connectionClosed(this);
+
+    // If an error triggered a server shutdown, act on it now
+    if (server._doQuit) {
+      server.stop();
+    }
+  },
+};
+
+function streamClosed(e) {
+  return e === Cr.NS_BASE_STREAM_CLOSED ||
+         (typeof e === "object" && e.result === Cr.NS_BASE_STREAM_CLOSED);
+}
+
+// Parse and dispatch incoming events from client
+function EventHandler(connection) {
+  this._connection = connection;
+
+  this._output = Cc["@mozilla.org/intl/converter-output-stream;1"]
+                   .createInstance(Ci.nsIConverterOutputStream);
+  this._output.init(this._connection.output, "UTF-8", 0, 0x0000);
+
+  this._input = Cc["@mozilla.org/intl/converter-input-stream;1"]
+                  .createInstance(Ci.nsIConverterInputStream);
+  this._input.init(this._connection.input, "UTF-8", 0,
+                   Ci.nsIConverterInputStream.DEFAULT_REPLACEMENT_CHARACTER);
+}
+EventHandler.prototype = {
+  // nsIInputStreamCallback
+  onInputStreamReady: function(input) {
+    DEBUG && debug("onInputStreamReady(input=" + input + ") on thread " +
+                   Services.tm.currentThread + " (main is " +
+                   Services.tm.mainThread + ")");
+
+    try {
+      let available = 0, numChars = 0, fullMessage = "";
+
+      // Read and concat messages from input stream buffer
+      do {
+        let partialMessage = {};
+
+        available = input.available();
+        numChars = this._input.readString(available, partialMessage);
+
+        fullMessage += partialMessage.value;
+      } while(numChars < available);
+
+      if (numChars != 0 && fullMessage.length > 0) { // While last readString contains something
+        // Handle concatenated JSON string
+        fullMessage = '[' + fullMessage.replace(/}{/g, '},{') + ']';
+
+        try {
+          // Parse JSON string to event objects
+          let events = JSON.parse(fullMessage);
+
+          events.forEach((event) => {
+            this._handleEvent(event);
+          });
+        } catch (e) {
+          DEBUG && debug ("Parse event error, drop this message");
+        }
+      }
+    } catch (e) {
+      if (streamClosed(e)) {
+        DEBUG && debug("WARNING: unexpected error when reading from socket; will " +
+                       "be treated as if the input stream had been closed");
+        DEBUG && debug("WARNING: actual error was: " + e);
+      }
+
+      // Input has been closed, but we're still expecting to read more data.
+      // available() will throw in this case, destroy the connection.
+      DEBUG && debug("onInputStreamReady called on a closed input, destroying connection");
+      this._connection.close();
+      return;
+    }
+
+    // Wait next message
+    input.asyncWait(this, 0, 0, Services.tm.currentThread);
+  },
+
+  // PRIVATE FUNCTIONS
+  _handleEvent: function(event) {
+    // TODO: Implement JPAKE pairing and control command dispatch here
+  },
+};
+
+RemoteControlService.init();
```

Comment:
`RemoteControlService.init()` does nothing but print a debug log. please remove it.